File: AuctionMarket/src/main/java/com/portfolio/auctionmarket/domain/auctions/service/AuctionService.java
package com.portfolio.auctionmarket.domain.auctions.service;

import com.portfolio.auctionmarket.domain.auctions.entity.Auction;
import com.portfolio.auctionmarket.domain.auctions.entity.AuctionStatus;
import com.portfolio.auctionmarket.domain.auctions.repository.AuctionRepository;
import com.portfolio.auctionmarket.domain.bids.entity.Bid;
import com.portfolio.auctionmarket.domain.bids.entity.BidStatus;
import com.portfolio.auctionmarket.domain.bids.repository.BidRepository;
import com.portfolio.auctionmarket.domain.orders.dto.OrderResponse;
import com.portfolio.auctionmarket.domain.orders.entity.Order;
import com.portfolio.auctionmarket.domain.orders.repository.OrderRepository;
import com.portfolio.auctionmarket.domain.products.entity.ProductStatus;
import com.portfolio.auctionmarket.domain.user.entity.User;
import com.portfolio.auctionmarket.domain.user.repository.UserRepository;
import com.portfolio.auctionmarket.global.error.CustomException;
import com.portfolio.auctionmarket.global.error.ErrorCode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuctionService {

    private final AuctionRepository auctionRepository;
    private final BidRepository bidRepository;
    private final OrderRepository orderRepository;
    private final UserRepository userRepository;

    @Transactional
    public Optional<OrderResponse> finishAuction(Long auctionId) {
        Auction auction = auctionRepository.findByIdWithPessimisticLock(auctionId)
                .orElseThrow(() -> new CustomException(ErrorCode.AUCTION_NOT_FOUND, "옥션을 찾을 수 없습니다."));

        User user = auction.getProduct().getSeller();

        if (auction.getStatus() != AuctionStatus.PROCEEDING) {
            return Optional.empty();
        }

        auction.changeStatus(AuctionStatus.ENDED);

        Optional<Bid> topBid = bidRepository.findTopByStatusAndAuctionOrderByBidIdDesc(BidStatus.ACTIVE, auction);

        if (topBid.isPresent()) {
            Bid winnerBid = topBid.get();
            Order order = Order.builder()
                .auction(auction)
                .buyer(winnerBid.getBidder())
                .finalPrice(winnerBid.getBidPrice())
                .build();
            orderRepository.save(order);
            user.addPoint(auction.getCurrentPrice());
            auction.getProduct().changeStatus(ProductStatus.SOLD);
            log.info("경매 낙찰 완료 - ID: {}", auctionId);
            return Optional.of(OrderResponse.from(order));
        } else {
            auction.getProduct().changeStatus(ProductStatus.FAILED);
            log.info("경매 유찰 완료 (입찰자 없음) - ID: {}", auctionId);
        }

        return Optional.empty();
    }
}

File: AuctionMarket/auction-app-user/src/main/kotlin/org/example/auctioncommon/domain/auctions/service/AuctionService.kt
package org.example.auctioncommon.domain.auctions.service

import org.example.auctioncommon.domain.auction.entity.Auction
import org.example.auctioncommon.domain.auction.entity.AuctionStatus
import org.example.auctioncommon.domain.auction.error.AuctionErrorCode
import org.example.auctioncommon.domain.auction.repository.AuctionRepository
import org.example.auctioncommon.domain.auction.service.AuctionProcessor
import org.example.auctioncommon.domain.bid.entity.Bid
import org.example.auctioncommon.domain.bid.entity.BidStatus
import org.example.auctioncommon.domain.bid.repository.BidRepository
import org.example.auctioncommon.domain.order.entity.Order
import org.example.auctioncommon.domain.order.repository.OrderRepository
import org.example.auctioncommon.domain.orders.dto.OrderResponse
import org.example.auctioncommon.domain.user.entity.User
import org.example.auctioncommon.global.error.CustomException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class AuctionService(
    private val auctionRepository: AuctionRepository,
    private val auctionProcessor: AuctionProcessor,
    private val bidRepository: BidRepository,
    private val orderRepository: OrderRepository,
) {


    @Transactional
    fun finishAuction(auctionId: Long): Optional<OrderResponse> {
        val auction: Auction = auctionRepository.findByIdWithPessimisticLock(auctionId)
            ?: throw CustomException(AuctionErrorCode.AUCTION_NOT_FOUND, "옥션을 찾을 수 없습니다.")

        val user: User? = auction.product?.seller

        auction.changeStatus(AuctionStatus.ENDED)

        val topBid: Optional<Bid> = Optional.of(bidRepository.findTopByStatusAndAuctionOrderByBidIdDesc(BidStatus.ACTIVE, auction) as Bid)

        val order: Order? = auctionProcessor.validateFinishAuction(auction, topBid, user)

        orderRepository.save(order)

        return Optional.empty()
    }
}
